### PR TITLE
Switched Java worker to non-blocking

### PR DESCRIPTION
**Pull request by berndruecker**
As discussed with Amara we see this as a quick win to non have too much blocking code right at the beginning

### CODE REVIEW

**Comment by akeller** [comment]
Any side effects we need to document here by not doing a `join()`, or no?

**Comment by menski** [comment]
If you don't join or check the result of the future at a later state you are not able to determine if your complete command was accepted by the broker, i.e. maybe you want to retry in this error case or discard the external outcome of the worker.

I don't see this as a blocker for this example code. But maybe not a good best practice for real applications, as they should do some kind of error handling. 

I don't know what you internally discussed so you came to the conclusion to remove it here.

**Comment by akeller** [comment]
Let me fill you in 😄 

There is a concern that the subset of the Zeebe community (so pre-Cloud GA) using Java does not understand async vs. sync or non-blocking vs. blocking. They copy the examples as a starting point and use them in production. This is problematic for exactly what you mentioned above - "it may not be a good best practice for real applications". 

@berndruecker considered this a quick win without having to provide a ton of education around non-blocking vs. blocking. 

While I don't think we should have to teach people best practices for Java programming, I do think we need to offer some sort of protection for folks who will use the examples in production without understanding the consequences.

**Comment by jwulf** [comment]
See this Forum thread: https://forum.camunda.io/t/how-to-know-that-an-exception-has-occurred-when-use-zeebeclient-in-this-way/2348

**Comment by berndruecker** [comment]
@jwulf Unfortunately, the forum says this thread does not exist?
 
@akeller I was thinking a bit more about this - and especially about "protection for people that copy examples into prod without fully understanding them". Which I like. Having this in mind I changed my mind and think we should add an 

`.exceptionally(throwable -> {throw new RuntimeException("Could not complete job: " + throwable.getMessage(), throwable);})`

in the code - then I consider that production safe enough.  While I still think that line might be confusing for some folks - it is clearly the best compromise if we expect that this code makes it to production more often than we think (scales but also handles the exception, further there is less risk of simply forgetting the "join").

So I think we can live with the fact that the code is slightly more complicated and could also add a short hint that you can also simply "join()" if you want to run it blocking with a link to best practices.

And then we observe in real life. WDYT?
Happy to adjust the PR

**Comment by menski** [comment]
I like your new proposal @berndruecker, and think it's a good compromise between async and still doing some error handling. My only comment is that as far as I know throwing an exception inside the `exceptionally` handler doesn't really raise the exception, it just breaks the future chaining but it's not transparent to the user. Maybe something like this would be okay?
```
 send().exceptionally(throwable -> {
            LOG.error("Failed to complete job", throwable);
            return null;
          });
```

**Comment by berndruecker** [comment]
Good point - the more important fact is how to write code after the future completes successfully.

I am wondering if the "log exception" approach is beneficial in real life (compared to not handling the exception at all) - as I think unhandled exceptions will be logged on error level by default (if I am not mistaken)?

**Comment by menski** [comment]
> the more important fact is how to write code after the future completes successfully.

Then maybe `whenComplete((result, error) -> {...})` would fit better?

> as I think unhandled exceptions will be logged on error level by default (if I am not mistaken)?

I'm no aware that this is how it works. As far as I know, as long as you are not requesting the result or attach a handler the future result is totally ignored.

**Comment by berndruecker** [comment]
I am also not completely sure. I just know that in my typically Spring Boot applications this exception is logged - but that might totally depend on loggers and configs (which opens up another rabbit whole we don't want to open).

I like the proposal of using "whenComplete" to show that this is there - any objections?

**Comment by berndruecker** [comment]
Added another commit as a proposal - WDYT?

If we keep that I would quickly also adjust the README

**Comment by menski** [comment]
lgtm :+1:

**Comment by menski** [comment]
BTW the spring worker also blocks right now, maybe we want to align it
https://github.com/camunda-cloud/camunda-cloud-get-started/blob/d315e0f0e469186b24c4985c0ff618f4da618f60/spring/src/main/java/io/camunda/getstarted/EmailWorker.java#L23

**Comment by berndruecker** [comment]
Totally right! I aligned it and also adjusted the readme and added a small hint

**Comment by falko** [comment]
+1

File: java/README.md [hunk 1]
```diff
@@ -108,3 +114,19 @@ mvn exec:java -P worker
...
+```
+which is a blocking call to wait for the issues command to be executed on the workflow engine. While this is very straightforward to use and produces easy-to-read code, blocking code is limited in terms of scalability. 
+
+That's why the worker showed a different pattern:
+```
+send().whenComplete((result, exception) -> {})
+```
+This registers a callback to be executed if the command on the workflow engines was executed or resulted in an exception. This allows for parallelism, which is especially interesting in workers. 
+
+This is discussed in more detial [this blog post about writing good workers for Camunda Cloud](https://blog.bernd-ruecker.com/writing-good-workers-for-camunda-cloud-61d322cad862).
```
Comment (akeller) [inline java/README.md:132]:
```suggestion
This is discussed in more detail in [this blog post about writing good workers for Camunda Cloud](https://blog.bernd-ruecker.com/writing-good-workers-for-camunda-cloud-61d322cad862).
```

File: java/README.md [hunk 2]
```diff
@@ -108,3 +114,19 @@ mvn exec:java -P worker
...
+```
+send().join()
+```
+which is a blocking call to wait for the issues command to be executed on the workflow engine. While this is very straightforward to use and produces easy-to-read code, blocking code is limited in terms of scalability. 
+
+That's why the worker showed a different pattern:
+```
+send().whenComplete((result, exception) -> {})
+```
+This registers a callback to be executed if the command on the workflow engines was executed or resulted in an exception. This allows for parallelism, which is especially interesting in workers.
```
Comment (akeller) [inline java/README.md:130]:
```suggestion
This registers a callback to be executed if the command on the workflow engine was executed or resulted in an exception. This allows for parallelism, which is especially interesting in workers. 
```

File: spring/README.md [hunk 3]
```diff
@@ -136,3 +143,20 @@ mvn spring-boot:run
...
+```
+send().join()
+```
+which is a blocking call to wait for the issues command to be executed on the workflow engine. While this is very straightforward to use and produces easy-to-read code, blocking code is limited in terms of scalability. 
+
+That's why the worker showed a different pattern:
+```
+send().whenComplete((result, exception) -> {})
+```
+This registers a callback to be executed if the command on the workflow engines was executed or resulted in an exception. This allows for parallelism, which is especially interesting in workers.
```
Comment (akeller) [inline spring/README.md:160]:
```suggestion
This registers a callback to be executed if the command on the workflow engine was executed or resulted in an exception. This allows for parallelism, which is especially interesting in workers. 
```

File: spring/README.md [hunk 4]
```diff
@@ -136,3 +143,20 @@ mvn spring-boot:run
...
+```
+which is a blocking call to wait for the issues command to be executed on the workflow engine. While this is very straightforward to use and produces easy-to-read code, blocking code is limited in terms of scalability. 
+
+That's why the worker showed a different pattern:
+```
+send().whenComplete((result, exception) -> {})
+```
+This registers a callback to be executed if the command on the workflow engines was executed or resulted in an exception. This allows for parallelism, which is especially interesting in workers. 
+
+This is discussed in more detial [this blog post about writing good workers for Camunda Cloud](https://blog.bernd-ruecker.com/writing-good-workers-for-camunda-cloud-61d322cad862).
```
Comment (akeller) [inline spring/README.md:162]:
```suggestion
This is discussed in more detail in [this blog post about writing good workers for Camunda Cloud](https://blog.bernd-ruecker.com/writing-good-workers-for-camunda-cloud-61d322cad862).
```

**Review by akeller — CHANGES_REQUESTED** [review]
Small typos

**Comment by berndruecker** [comment]
Great - thanks! Applied the fixes

**Comment by berndruecker** [comment]
@akeller Jut double-checking - anything left to do for me?

**Review by akeller — APPROVED** [review]

**Comment by akeller** [comment]
Should be good! I'll squash and merge, but @menski will likely need to do a release :)

**Pull request merged**